fees/incrementor.go

```diff
@@ -0,0 +1,321 @@
+	return &GasPriceIncremenetor{
+		pullInterval: pullInterval,
+
+		storage: storage,
+		bc:      cl,
+		sign:    signer,
+
+		syncer: newSyncer(),
+		stop:   make(chan struct{}, 0),
+	}
+}
+
+// Run starts the gas price incrementer.
+//
+// It will query the given storage for any entries that it needs to check
+// for gas increase, trying to check their status.
+func (i *GasPriceIncremenetor) Run() {
+	process := func(txs []Transaction) {
+		for _, tx := range txs {
+			switch tx.State {
+			case TxStateFailed, TxStateSucceed:
+				// Force skip transactions that are finalyzed.
+			default:
+				i.tryWatch(tx)
+			}
+		}
+	}
+
+	for {
+		select {
+		case <-i.stop:
+			return
+
+		case <-time.After(i.pullInterval):
+			txs, err := i.storage.GetIncrementorTransactionsToCheck()
+			if err != nil {
+				continue
+			}
+			process(txs)
+		}
+	}
+}
+
+// AttachLogFunc attaches a new log func incremeter thread.
+// The given log func is called every time the running incrementer thread
+// created by the `Run` method encouters an error.
+//
+// This method is not thread safe and should be called before Run.
+func (i *GasPriceIncremenetor) AttachLogFunc(logFn LogFunc) {
+	i.logFn = logFn
+}
+
+// Stop stops the execution of GasPriceIncrementer thread created by the Run method.
+func (i *GasPriceIncremenetor) Stop() {
+	i.once.Do(func() {
+		close(i.stop)
+	})
+}
+
+// InsertInitial uses the given storage to insert an new transaction which
+// will later be retreived using `GetTransactionsToCheck` in order to check
+// it's state and retry with higher gas price if needed.
+func (i *GasPriceIncremenetor) InsertInitial(tx *types.Transaction, opts TransactionOpts, chainID int64) error {
+	if err := opts.validate(); err != nil {
+		return fmt.Errorf("invalid opts given: %w", err)
+	}
+	newTx, err := newTransaction(tx, chainID, opts)
+	if err != nil {
+		return fmt.Errorf("failed to create new transaction: %w", err)
+	}
+
+	return i.storage.UpsertIncrementorTransaction(*newTx)
+}
+
+// tryWatch will try to watch a transaction.
+// If a transaction is already being watched, it will get skipped.
+func (i *GasPriceIncremenetor) tryWatch(tx Transaction) {
+	if i.syncer.txBeingWatched(tx) {
+		// Already watching
+		return
+	}
+	if err := tx.Opts.validate(); err != nil {
+		i.log(tx, fmt.Errorf("can't increment gas price, got wrong tx opts: %w", err))
+		return
+	}
+
+	i.syncer.txMarkBeingWatched(tx)
+	go func() {
+		defer i.syncer.txRemoveWatched(tx)
+		if err := i.watchAndIncrement(tx); err != nil {
+			i.log(tx, err)
+		}
+	}()
+}
+
+func (i *GasPriceIncremenetor) watchAndIncrement(tx Transaction) error {
+	timeout := time.After(tx.Opts.Timeout)
+	incTimer := time.NewTicker(tx.Opts.IncreaseInterval)
+	defer incTimer.Stop()
+
+	checkTimer := time.NewTicker(tx.Opts.CheckInterval)
+	defer checkTimer.Stop()
+
+	for {
+		select {
+		case <-i.stop:
+			return nil
+		case <-checkTimer.C:
+			receipt, err := i.getReceipt(tx)
```

Comment:
i'd exctract functions from these huge code blocks in the cases, it would make reading and comprehending a bit easier.